File: pkg/simecolModels/R/OmexDia.r
################################################################################
#
# OmexDia - a model of Carbon, Nitrogen and oxygen diagenesis in (marine) sediments
# Soetaert et al., 1996.
#
# Model describing the dynamics of biogenic and dissolved silicate
# in a marine sediment. (Soetaert and Herman, 2008).
#
# the Newton-raphson method is used to solve for the steady-state condition
# Solved with rootSolve
#
################################################################################

OmexDia <- function() {
  new("odeModel",
    main = function(time = 0, state, parms) {
      with (c(as.list(parms),inputs$boxes), {

Flux  <- MeanFlux * (1+sin(2*pi*time/365))

FDET  <- state[1:N]
SDET  <- state[(N+1)  :(2*N)]
O2    <- state[(2*N+1):(3*N)]
NO3   <- state[(3*N+1):(4*N)]
NH3   <- state[(4*N+1):(5*N)]
ODU   <- state[(5*N+1):(6*N)]

# Transport fluxes
#==================

# Solid substances: upper boundary = deposition, lower=zero-gradient
FDETFlux<- -Db*(1-IntPor)*diff(c(FDET[1],FDET,FDET[N]))/thick   + #diffusion
            w * c(FDET[1],FDET)                                   #advection
FDETFlux[1] <-  Flux*pFast                                        #deposition

SDETFlux<- -Db*(1-IntPor)*diff(c(SDET[1],SDET,SDET[N]))/thick   +
            w * c(SDET[1],SDET)
SDETFlux[1] <-  Flux*(1.-pFast)

# Solute substances: upper boundary=imposed concentration
O2Flux   <- -DispO2  *   IntPor *diff(c(bwO2  ,O2 ,O2[N] ))/thick    +
            w * c(bwO2 ,O2)

NO3Flux  <- -DispNO3 *   IntPor *diff(c(bwNO3 ,NO3,NO3[N]))/thick    +
            w * c(bwNO3 ,NO3)

NH3Flux  <- -DispNH3 *   IntPor *diff(c(bwNH3 ,NH3,NH3[N]))/thick    +
            w * c(bwNH3 ,NH3)

ODUFlux  <- -DispODU *   IntPor *diff(c(bwODU ,ODU,ODU[N]))/thick    +
            w * c(bwODU ,ODU)


# production of DIC and DIN, expressed per cm3 LIQUID/day
DICprod_Min <- (rFast*FDET         +rSlow*SDET       )*(1.-Porosity)/Porosity
DINprod_Min <- (rFast*FDET*NCrFdet+rSlow*SDET*NCrSdet)*(1.-Porosity)/Porosity

# oxic mineralisation, denitrification, anoxic mineralisation
Oxicminlim <- O2/(O2+ksO2oxic)                     # limitation terms
Denitrilim <- (1-O2/(O2+kinO2denit))*NO3/(NO3+ksNO3denit)
Anoxiclim  <- (1-O2/(O2+kinO2anox))*(1-NO3/(NO3+kinNO3anox))
Rescale    <- 1/(Oxicminlim+Denitrilim+Anoxiclim)

OxicMin    <- DICprod_Min*Oxicminlim*Rescale        # oxic mineralisation
Denitrific <- DICprod_Min*Denitrilim*Rescale        # Denitrification
AnoxicMin  <- DICprod_Min*Anoxiclim *Rescale        # anoxic mineralisation

# reoxidation and ODU deposition
Nitri      <- rnit  *NH3*O2/(O2+ksO2nitri)
OduOx      <- rODUox*ODU*O2/(O2+ksO2oduox)

pDepo      <- min(1,0.233*(w*365)^0.336 )
OduDepo    <- AnoxicMin*pDepo

# the rate of change=- Flux gradient    + biogeochemistry
dFDET <- -diff(FDETFlux)/thick/(1-Porosity) - rFast*FDET
dSDET <- -diff(SDETFlux)/thick/(1-Porosity) - rSlow*SDET
dO2   <- -diff(O2Flux) /thick/Porosity -  OxicMin      -2* Nitri -      OduOx
dNH3  <- -diff(NH3Flux)/thick/Porosity + (DINprod_Min  - Nitri) / (1.+NH3Ads)
dNO3  <- -diff(NO3Flux)/thick/Porosity - 0.8*Denitrific + Nitri
dODU  <- -diff(ODUFlux)/thick/Porosity + AnoxicMin  - OduOx - OduDepo

#

return(list(c(dFDET,dSDET,dO2,dNO3,dNH3,dODU),
            Cflux = Flux, O2flux=O2Flux[1],NH3flux=NH3Flux[1],
            NO3flux=NO3Flux[1],ODUflux=ODUFlux[1]))

       })
    },
    parms = c(
     # sediment parameters
# N should NOT be a parameter; it shoudl be calculated, but i cannot make the
# model work unless N is a parameter...
     N        = 300,        # Total number of boxes
     thick    = 0.05,       # thickness of sediment layers (cm)
     por0     = 0.9,        # surface porosity (-)
     pordeep  = 0.7,        # deep porosity    (-)
     porcoef  = 2  ,        # porosity decay coefficient  (/cm)

     dB0      = 1/365,      # cm2/day       - bioturbation coefficient
     dBcoeff  = 2    ,
     mixdepth = 5    ,      # cm

     MeanFlux = 20000/12*100/365,  # nmol/cm2/d - Carbon deposition: 20gC/m2/yr
     rFast    = 0.01            ,  #/day        - decay rate fast decay detritus
     rSlow    = 0.00001         ,  #/day        - decay rate slow decay detritus
     pFast    = 0.9             ,  #-           - fraction fast detritus in flux
     w        = 0.1/1000/365    ,  # cm/d       - advection rate
     NCrFdet  = 0.16            ,  # molN/molC  - NC ratio fast decay detritus
     NCrSdet  = 0.13            ,  # molN/molC  - NC ratio slow decay detritus

# Nutrient bottom water conditions
     bwO2     = 300             ,  #mmol/m3     Oxygen conc in bottom water
     bwNO3    = 10              ,  #mmol/m3
     bwNH3    = 1               ,  #mmol/m3
     bwODU    = 0               ,  #mmol/m3

# Nutrient parameters
     NH3Ads     = 1.3        ,  #-           Adsorption coeff ammonium
     rnit       = 20.        ,  #/d          Max nitrification rate
     ksO2nitri  = 1.         ,  #umolO2/m3   half-sat O2 in nitrification
     rODUox     = 20.        ,  #/d          Max rate oxidation of ODU
     ksO2oduox  = 1.         ,  #mmolO2/m3   half-sat O2 in oxidation of ODU
     ksO2oxic   = 3.         ,  #mmolO2/m3   half-sat O2 in oxic mineralisation
     ksNO3denit = 30.        ,  #mmolNO3/m3  half-sat NO3 in denitrification
     kinO2denit = 1.         ,  #mmolO2/m3   half-sat O2 inhib denitrification
     kinNO3anox = 1.         ,  #mmolNO3/m3  half-sat NO3 inhib anoxic degr
     kinO2anox  = 1.         ,  #mmolO2/m3   half-sat O2 inhib anoxic min

# Diffusion coefficients, temp = 10dgC
     Temp       = 10         ,               # temperature
     DispO2     = 0.955    +10*0.0386 ,    #a+ temp*b
     DispNO3    = 0.844992 +10*0.0336 ,
     DispNH3    = 0.84672  +10*0.0336 ,
     DispODU    = 0.8424   +10*0.0242

    ),
    times  = 0,                     # t=0 for steady-state calculation
    initfunc = function(obj) {      # initialisation
      pars <- parms(obj)
      with(as.list(pars),{
# THIS DID NOT WORK
#        Intdepth = seq(0,10,by=thick)         # depth at upper interface of each layer
        Nint     = N+1                             # number of interfaces
        Intdepth = seq(0,by=thick,len=Nint)        # depth at upper interface of each layer

        Depth    = 0.5*(Intdepth[-Nint] +Intdepth[-1]) # depth at middle of each layer
# THOMAS: N should be estimated here, but then I cannot access N in the solver
# so I had to make N a parameter can this be fixed?....
#       N        = length(Depth)                       # number of layers

        init(obj) = rep(1, 6 * N)                      # initial condition; any positive number will do
        Porosity = pordeep + (por0-pordeep)*exp(-Depth*porcoef)     # porosity profile, middle of layers
        IntPor   = pordeep + (por0-pordeep)*exp(-Intdepth*porcoef)  # porosity profile, upper interface

        Db       = pmin(dB0,dB0*exp(-(Intdepth-mixdepth)*dBcoeff))  # Bioturbation profile
        inputs(obj)$boxes <- list(Depth=Depth,Intdepth=Intdepth,
                                  Porosity=Porosity,IntPor=IntPor,Db=Db)
        return(obj)
      })
    },
    # this model comes with a user defined solver,
    #   i.e. a function instead of a character that points to an existing solver
    solver = function(y, times, func, parms, ...) {
      # steady-state condition of state variables, one vector
      out <- steady.1D(y, time=times, func, parms, nspec=6, pos=TRUE, ...)$y
      # rearrange as data.frame
      with (as.list(parms),
        data.frame(FDET = out[1:N          ],
                   SDET = out[(N+1)  :(2*N)],
                   O2   = out[(2*N+1):(3*N)],
                   NO3  = out[(3*N+1):(4*N)],
                   NH3  = out[(4*N+1):(5*N)],
                   ODU  = out[(5*N+1):(6*N)])
      )
    }
  )
}

File: pkg/simecolModels/src/omexdia.f
!##########################################################################
! OMEXDIA - Model describing the dynamics of carbon, nitrogen and oxygen
! in a marine sediment. (Soetaert et al., 1996 a,b).
!
! Soetaert K, PMJ Herman and JJ Middelburg, 1996.
! A model of early diagenetic processes from the shelf to abyssal depths.
! Geochim. Cosmochim. Acta, 60(6):1019-1040.
! Soetaert K, PMJ Herman and JJ Middelburg, 1996b.
! Dynamic response of deep-sea sediments to seasonal variation: a model.
! Limnol. Oceanogr. 41(8): 1651-1668.
!
! The model describes six state variables, in 200 layers:
!
! fast decaying organic carbon (FDET), solid substance (nmol/cm3 solid)
! slow decaying organic carbon (SDET), solid substance
! Oxygen (O2), dissolved substance (nmol/cm3 liquid)
! Nitrate (NO3), dissolved substance
! Ammonium (NH3), dissolved substance
! Oxygen demand unit (ODU), dissolved,
!                    =lump sum of reduced substances other than ammonium
!##########################################################################


!==========================================================================
!==========================================================================
! initialisation subroutine: 
! initialises the common block with 25 parameter values, 
! followed by thicknesses, porosities, bioturbation values
!==========================================================================
!==========================================================================

       subroutine initomexdia (steadyparms)
       external steadyparms
       integer,parameter :: N=200
       integer,parameter :: nc = 2*N + 3*(N+1) + 25

       double precision parms(nc)
       common /myparms/parms

       call steadyparms(nc, parms)

       return
       end


!==========================================================================
!==========================================================================
! subroutine calculating the rate of change
! the omexdia model
!==========================================================================
!==========================================================================

      subroutine omexdiamod (neq, t, Conc, dConc, yout, ip)
      
      implicit none


!......................... declaration section.............................
      integer           :: neq, ip(*), i
      integer,parameter :: N=200         
      double precision  :: t, Conc(*), dConc(*), yout(*), Flux(N+1)
      double precision :: por(N),intpor(N+1),Db(N+1),dx(N),dxInt(N+1)
      double precision :: Fdet(N),Sdet(N),O2(N),NO3(N),NH3(N),ODU(N)
      double precision :: dFdet(N),dSdet(N),dO2(N),dNO3(N),               &
     &                     dNH3(N),dODU(N)

      double precision :: cflux,Cprod(N),Nprod(N),Rescale(N)
      double precision :: Oxicminlim(N),Denitrilim(N),Anoxiclim(N) 
      double precision :: Oxicmin(N),Denitrific(N),anoxicmin(N)
      double precision :: nitri(N),oduox(N),odudepo(N)
      double precision :: pdepo

       
      double precision :: MeanFlux,rFast,rSlow ,pFast,w,NCrFdet,          & 
     &  NCrSdet,bwO2,bwNO3,bwNH3,bwODU,NH3Ads,rnit,ksO2nitri,             &
     &  rODUox,ksO2oduox,ksO2oxic,ksNO3denit,kinO2denit,kinNO3anox,       &
     &  kinO2anox,DispO2,DispNO3,DispNH3,DispODU   

      common /myparms    /MeanFlux,rFast,rSlow ,pFast,w,NCrFdet,          & 
     &  NCrSdet,bwO2,bwNO3,bwNH3,bwODU,NH3Ads,rnit,ksO2nitri,             &
     &  rODUox,ksO2oduox,ksO2oxic,ksNO3denit,kinO2denit,kinNO3anox,       &
     &  kinO2anox,DispO2,DispNO3,DispNH3,DispODU,dx,dxint,                &
     &  por,intpor,Db   

      double precision :: integrate
!............................ statements ..................................
       if (IP(1) < 9)  call rexit("nout should be at least 9")

! from Conc to fdet, sdet, o2,...
       do i = 1,N 
          Fdet(i) = Conc(i)
          Sdet(i) = Conc(i+N)          
          O2  (i) = Conc(i+2*N)          
          NO3 (i) = Conc(i+3*N)                              
          NH3 (i) = Conc(i+4*N)          
          ODU (i) = Conc(i+5*N)          
       enddo

      CFlux  = MeanFlux * (1+sin(2.*3.14158*t/365.))
      yout(1) = CFlux
	 
! Rate of change due to transport: solid substances
      CALL tran1Dsol(FDET,cFlux*pFast,Db,w,por,intpor,dx,dxint,            &
     &               Flux,dFDET)

      CALL tran1Dsol(SDET,cFlux*(1.-pFast),Db,w,por,intpor,dx,dxint,       &
     &               Flux,dSDET)

! Rate of change due to transport: solute substances
      CALL tran1dliq(O2 ,bwO2 ,dispO2 ,w,por,intpor,dx,dxint,Flux,dO2)     
      yout(2) = Flux(1) ! sediment-water O2 flux

      CALL tran1dliq(NH3,bwNH3,dispNH3/(1+NH3Ads),w,por,intpor,dx,dxint,   &
     &               Flux,dNH3)                        
      yout(3) = Flux(1)

      CALL tran1dliq(NO3,bwNO3,dispNO3,w,por,intpor,dx,dxint,Flux,dNO3)                        
      yout(4) = Flux(1)

      CALL tran1dliq(ODU,bwODU,dispODU,w,por,intpor,dx,dxint,Flux,dODU)                        
      yout(5) = Flux(1)

! Production of DIC and DIN, expressed per cm3 LIQUID/day
      Cprod= (rFast*FDET        +rSlow*SDET        )*(1.-por)/por
      Nprod= (rFast*FDET*NCrFdet+rSlow*SDET*NCrSdet)*(1.-por)/por

! Oxic mineralisation, denitrification, anoxic mineralisation
! first the limitation terms
      Oxicminlim = O2/(O2+ksO2oxic)                ! limitation terms
      Denitrilim = (1.-O2/(O2+kinO2denit))*NO3/(NO3+ksNO3denit)
      Anoxiclim  = (1.-O2/(O2+kinO2anox))*(1-NO3/(NO3+kinNO3anox))
      Rescale    = 1./(Oxicminlim+Denitrilim+Anoxiclim)

! then the mineralisation rates
      OxicMin    = Cprod*Oxicminlim*Rescale        ! oxic mineralisation
      Denitrific = Cprod*Denitrilim*Rescale        ! Denitrification
      AnoxicMin  = Cprod*Anoxiclim *Rescale        ! anoxic mineralisation

! reoxidation and ODU deposition
      Nitri      = rnit  *NH3*O2/(O2+ksO2nitri)
      OduOx      = rODUox*ODU*O2/(O2+ksO2oduox)

      pDepo      = min(1.d0,0.233*(w*365)**0.336 )
      OduDepo    = AnoxicMin*pDepo 

! Update the rate of change with biogeochemical processes
      dFDET = dFDET  - rFast*FDET
      dSDET = dSDET  - rSlow*SDET
      dO2   = dO2    - OxicMin        -2* Nitri -      OduOx
      dNH3  = dNH3   + (Nprod           - Nitri) / (1.+NH3Ads)
      dNO3  = dNO3   - 0.8*Denitrific   + Nitri
      dODU  = dODU   + AnoxicMin                - OduOx - OduDepo

! from dfdet, dsdet, do2,... to dconc
       do i = 1,N
          dConc(i)   = dfdet(i) 
          dConc(i+N) = dsdet(i) 
          dConc(i+2*N) = dO2(i) 
          dConc(i+3*N) = dNO3(i) 
          dConc(i+4*N) = dNH3(i) 
          dConc(i+5*N) = dODU(i) 
       enddo

! the integrated rates
! Total mineralisation
      yout(6) = integrate(Cprod,dx,por,N)
! Total oxic mineralisation
      yout(7) = integrate(OxicMin,dx,por,N)
! Total denitrificatin
      yout(8) = integrate(Denitrific,dx,por,N)
! Total nitrification
      yout(9) = integrate(Nitri,dx,por,N)
      return
      end

!==========================================================================
! integrated quantities;
! solid substance y is in mmol/ m3 solid
! liquid substance y is in mmol /m3 liquid.
! converted to mmol/m3 bulk (* 1-por) or *(por) and integrate; porfac is
!==========================================================================

      double precision function integrate(y,dx,porfac,N)
      integer          :: i,N
      double precision :: y(N), dx(N), porfac(N)

      integrate = 0.d0
      do i = 1, N
       integrate = integrate + y(I)*dx(I)*porfac(I)
      enddo       

      end
       
!==========================================================================
! transport of liquids
!==========================================================================

       subroutine tran1dliq(y,yup,disp,v,por,porint,dx,dxint,Flux,dy)

       integer,parameter :: N=200         
       double precision, intent(in):: y(N),disp,v,por(N),porint(N+1)
       double precision, intent(in):: dx(N),dxint(N+1),yup 
       double precision, intent(out):: flux(N+1),dy(N)
       integer :: i
       double precision :: ydown,porfac(N+1)

! boundaries
       ydown = y(N)       

! advection, corrected for porosity
         do i = 1,N+1
           porfac(i) = porint(N+1)/porint(i)
         ENDDO

! dispersion flux  
         do i =1,N-1
           Flux(i+1) = (y(i)-y(i+1))/dxint(i+1)
         enddo
         Flux(1)     = (yup-y(1))/dxint(1)  
         Flux(N+1)   = (y(N)-ydown)/dxint(N+1)
         do i =1,N+1
          Flux(i) = Disp*Flux(i)
         enddo 

! advection flux in direction of axis - centered differences
         do i =2,N+1
           Flux(i) = Flux(i) + 0.5*v*porfac(i)*y(i-1)
         enddo
         Flux(1) = Flux(1) + 0.5*v*porfac(1)*yup

         do i =1,N
           Flux(i) = Flux(i)   + 0.5*v*porfac(i)*y(i)
         enddo
         Flux(N+1) = Flux(N+1) + 0.5*v*porfac(N+1)*ydown

! Fluxes per bulk
       Flux = Flux * porint

! rate of change = flux gradient, corrected for surface (porosity) changes 
       do i =1,N
         dy(i) = (Flux(i)-Flux(i+1))  /por(i)/dx(i)
       enddo
       
       end subroutine

!==========================================================================
! transport of solids
!==========================================================================

       subroutine tran1dsol(y,fluxup,disp,v,por,porint,dx,dxint,Flux,dy)

       integer,parameter :: N=200         
       double precision, intent(in):: y(N),fluxup,disp(N+1),v,por(N)
       double precision, intent(in):: dx(N),dxint(N+1),porint(N+1)
       double precision, intent(out):: flux(N+1),dy(N)
       integer :: i
       double precision :: yup,ydown,porfac(N+1)
       character(len=80) msg

! boundaries
       yup   = y(1)       
       ydown = y(N)       

! dispersion flux  
         do i =1,N-1
           Flux(i+1) = (y(i)-y(i+1))/dxint(i+1)
         enddo

         Flux(N+1)   = (y(N)- ydown)/dxint(N+1)
         do i =1,N+1
          Flux(i) = Disp(i)*Flux(i)
         enddo 
! advection, corrected for porosity
         do i = 1,N+1
           porfac(i) = (1.-porint(N+1))/(1.-porint(i))
         ENDDO
! advection flux in direction of axis
         do i =2,N+1
           Flux(i) = Flux(i) + 0.5*v*porfac(i)*y(i-1)
         enddo

         do i =1,N
           Flux(i) = Flux(i)     + (1.-0.5)*v*porfac(i)*y(i)
         enddo
         Flux(N+1) = Flux(N+1) + (1.-0.5)*v*porfac(N+1)*ydown

! fluxes per bulk
       Flux = Flux * (1-porint)

! upper boundary flux
       Flux(1) = fluxup
              
! rate of change = flux gradient, corrected for surface (porosity) changes 
       do i =1,N
         dy(i) = (Flux(i)-Flux(i+1)) /(1.-por(i))/dx(i)         
       enddo
       
       end subroutine
       
